lib/probes/mongodb.js

```diff
@@ -0,0 +1,348 @@
+    if (typeof opts === 'function') {
+      cb = opts
+      opts = {}
+    }
+    return ao.instrument(
+      () => {
+        return {
+          name: 'mongodb',
+          kvpairs: makeData(this, ns, cmd)
+        }
+      },
+      done => handler.call(this, ns, cmd, opts, done),
+      conf,
+      cb
+    )
+  })
+}
+
+function patchCommands (obj) {
+  makeWrapper(obj, 'insert', (data, cmd) => {
+    data.Insert_Document = JSON.stringify(cmd)
+  })
+  makeWrapper(obj, 'update', (data, cmd) => {
+    data.Query = JSON.stringify(cmd.map(getQuery))
+    data.Update_Document = JSON.stringify(cmd.map(getUpdate))
+  })
+  makeWrapper(obj, 'remove', (data, cmd) => {
+    data.Query = JSON.stringify(cmd.map(getQuery))
+  })
+
+  patchCommand(obj)
+}
+
+function patchCursor (cursor) {
+  if (typeof cursor.next !== 'function') {
+    logMissing('cursor.next()')
+    return
+  }
+  shimmer.wrap(cursor, 'next', handler => function (cb) {
+    const self = this
+    let span
+
+    return ao.instrument(
+      () => {
+        return {
+          name: 'mongodb',
+          kvpairs: makeData(this.topology, this.ns, this.cmd)
+        }
+      },
+      done => handler.call(this, function () {
+        if (span) span.events.exit.set({
+          CursorId: self.cursorState.cursorId.toString(),
+          CursorOp: 'next'
+        })
+        return done.apply(this, arguments)
+      }),
+      conf,
+      cb
+    )
+  })
+}
+
+//
+// Helpers
+//
+
+// List deconstructors
+function getQuery (v) {return v.q || v.query}
+function getUpdate (v) {return v.u || v.update}
+
+// Command identifiers
+function ifDef (...args) {
+  return function (o) {
+    for (let i = 0; i < args.length; i++) {
+      if (args[i] in o || args[i].toLowerCase() in o) {
+        return true;
+      }
+    }
+    return false;
+    //return args.reduce((r, v) => r || v in o || v.toLowerCase() in o, false)
+  }
+}
+
+// different versions of mongodb and mongodb-core have variations in naming
+// and object organization.
+function serverDetails ({s = {}}) {
+  let o;
+  if (s.serverDetails) {
+    o = s.serverDetails;
+  } else if (s.replicaSetState && s.replicaSetState.primary) {
+    o = s.replicaSetState.primary;
+    // eslint-disable-next-line max-len
+  } else if (s.coreTopology && s.coreTopology.s && s.coreTopology.s.replicaSetState && s.coreTopology.s.replicaSetState.primary) {
+    o = s.coreTopology.s.replicaSetState.primary;
+  } else if (s.replState && s.replState.primary) {
+    o = s.replState.primary.s.serverDetails;
+  } else {
+    o = {name: s.options.host + ':' + (s.options.port || 27017)};
+  }
+  return o;
+}
+
+function makeBaseData (ctx, ns) {
+  // for some reason cursors don't use a namespace object.
+  if (typeof ns === 'string') {
+    // a collection name can contain a '.' so don't just split.
+    const dot = ns.indexOf('.');
+    const db = ns.slice(0, dot);
+    const collection = ns.slice(dot + 1);
+    ns = {db, collection};
+  }
```

Comment:
if ns is not a string is it an object that has 'db' and 'collection' keys for sure?

Reply:
it's hard to tell with mongo but i've seen nothing else. it needs to have a db and collection value. i had changed everything to expect the namespace object but it turned out that a couple api calls still used a ns string. prior to this namespace was always a string.